Add destinations page

stveit opened the pull request:
Fixes Uninett/Argus#1001 

Replaced by Uninett/Argus#970

Adds page where you can see and edit existing destinations and create new ones.

Since we are kinda locked to using the serializer for validating destinations (for now, changes will be made at some point) I ended up baking the serializer into the forms for creating/updating destinations. Its a bit wonky but should work fine until we can move on from the serializer and have `ModelForm` do everything they way it is supposed to.

Needs more styling to be complete, but not sure how to do it properly.. Error in fields are now added to the right of the field the error belongs to, so the buttons and everything gets moved to the right when theres an error.

hmpf commented:
Lookswise:

Having "Label", "Media", "Value" inside the same box as the actual changeable value looks terrible. I think the login form has the value on the outside.

There should be the rounded box around each if we want to have boxes, again see existing code.

I'm not sure the dropdowns are needed at all. We should not expect a zillion destinations per user.

The grey behind each box has to go. Too low contrast.

The "media" input in each box is redundant.

I think a header would look nicer than having one box for each type.

hmpf commented:
Functions wise: make it work without HTMx first. Django has something called "formsets" you can use, see the abandoned code in #109 (which tries to do too much at once as well).

hmpf commented:
DELETE and PATCH is API-stuff and htmx-stuff, not plain HTML. Make it work with plain HTML first.

hmpf commented on src/argus_htmx/destinations/forms.py:
```diff
@@ -0,0 +1,9 @@
+from argus.notificationprofile.models import Media
+
+
+class DestinationForm(forms.Form):
```
Could probably be a ModelForm.

hmpf commented on src/argus_htmx/destinations/urls.py:
```diff
@@ -1,21 +1,11 @@
 app_name = "htmx"
 urlpatterns = [
-    path("", placeholder, name="destination-placeholder"),
+    path("", views.destinations, name="destinations"),
+    path("create_destination/", views.destinations_create, name="destinations-create"),
+    path("update_destination/<int:pk>/", views.destinations_update, name="destinations-update"),
+    path("delete_destination/<int:pk>/", views.destinations_delete, name="destinations-delete"),
```
POST to /destination/: create

POST to /destination/<int:pk>/: update

POST to /destination/<int:pk>/: delete

This is the old, well tested way.

stveit replied:
Made it like this, so now /destination/ is used to get the site itself with GET and POST to create a destination

hmpf commented on src/argus_htmx/destinations/views.py:
```diff
@@ -0,0 +1,81 @@
+    return render(request, "htmx/destinations/destinations.html", context=context)
+
+
+def _get_destinations_and_forms_grouped_by_media(user) -> dict[Media, list[DestinationConfig]]:
+    """Returns dict where key is media and value is list of tuples
+    containing a destination and a pre-filled form for that destination."""
+    grouped_destinations = {}
+
+    media = Media.objects.all()
+    for m in media:
+        grouped_destinations[m] = []
+
+    destinations = user.destinations.all()
+    for destination in destinations:
+        form = DestinationForm(
+            initial={
+                "label": destination.label,
+                "media": destination.media.slug,
+                "value": destination.settings.get("email_address", ""),
+            }
+        )
+        form.fields["media"].widget.attrs["hidden"] = True
+        grouped_destinations[destination.media].append((destination, form))
+
+    return grouped_destinations
```
Not needed with ModelForm. You still might want to group though: sort by type, then label.

hmpf reviewed (commented):
I think this is a good candidate for class based views actually. We should have a meeting.

hmpf commented on src/argus_htmx/destinations/views.py:
```diff
@@ -1,81 +1,128 @@
-        destination.save()
+        serializer.is_valid(raise_exception=True)
+        serializer.save(user=request.user)
+
```
I was thinking you'd use the form to save, not go via a serializer. There's magic needed but we all ought to know that magic.

hmpf commented on src/argus_htmx/destinations/views.py:
```diff
@@ -1,81 +1,128 @@
-        destination.label = form.cleaned_data.get("label", "")
-        destination.settings = {"email_address": form.cleaned_data["value"]}
-        destination.save()
+        settings_key = _get_settings_key_for_media(destination.media)
```
You don't need line 84.

stveit committed:
Add view for destinations page

stveit committed:
Add form for creating a destination

stveit committed:
Add view for creating a destination

stveit committed:
Add view for deleting a destination

Pretty much copied from the destinations API, needs some changes
so it wont just crash to 500 when it raises a ValidationError

This commit doesnt make it show in the frontend, the button for
deleting a destination will be grouped with the forms for
updating destinations

stveit committed:
Add form for updating a destination

stveit committed:
Add view for updating a destination

stveit committed:
Add subtemplate for rendering update/delete form

stveit committed:
Add destinations link to navbar dropdown

stveit committed:
Update css

Seems like a lot of changes .. but this is what the
wailwindcss script generated

stveit commented on src/argus_htmx/destinations/forms.py:
```diff
@@ -0,0 +1,79 @@
+            if settings := self.serializer.validated_data.get("settings"):
+                self.cleaned_data["settings"] = settings
+
+        if label := self.cleaned_data["label"]:
+            filter = DestinationConfig.objects.filter(label=label)
+            if self.instance:
+                filter = filter.exclude(pk=self.instance.pk)
+            if filter.exists():
+                self.add_error("label", "Name must be unique per media")
```
If label is `""` then this wont work

stveit replied:
But it might be fine? `""` seems to get translated to None when the object is saved to the database, and the constraints seem to allow multiple destinations without a label.

But the thing that should be changes is `self.cleaned_data["label"] `should be `self.cleaned_data.get("label")`. This doesnt cause any issues when testing, because the frontend sends an empty string for label if the input box is empty, but since this form doesnt actually require label to be set, we should assume that there are situations where it matters.
```suggestion
        if label := self.cleaned_data.get("label"):
            filter = DestinationConfig.objects.filter(label=label)
            if self.instance:
                filter = filter.exclude(pk=self.instance.pk)
            if filter.exists():
                self.add_error("label", "Name must be unique per media")
```

stveit replied:
Additionally, it might be wrong to check the uniqueness of the name here in the form validator. There is already a constraint in the DestinationConfig model for exactly this. This could then be handled by trying to save and see if the constraint throws an exception and take it from there. This is more or less a repetition of that constraint. Buut you would need some code anyway, and it might be easier to just let the form handle everything, instead of having to catch the constraint exception and cram errors into the form from the outside

stveit commented on src/argus_htmx/destinations/forms.py:
```diff
@@ -0,0 +1,79 @@
+    def clean(self):
+        super().clean()
+        settings_key = _get_settings_key_for_media(self.cleaned_data["media"])
+        self.cleaned_data["settings"] = {settings_key: self.cleaned_data["settings"]}
```
Add comment to explain that this is for making the value compatible with JSONField, which is what DestinationConfig.settings should really be. The form overrides it to be a CharField for the frotnend to make more sense (we shouldnt expect the user to give a full valid dict in JSON format as input, only the email address / phone number itself should be requires)

stveit commented on src/argus_htmx/destinations/forms.py:
```diff
@@ -0,0 +1,79 @@
+        media = self.cleaned_data["media"]
+        settings_key = _get_settings_key_for_media(media)
+
+        # Add error messages from serializer to form
+        if not self.serializer.is_valid():
+            for error_name, error_detail in self.serializer.errors.items():
+                if error_name in ["media", "label", settings_key]:
+                    if error_name == settings_key:
+                        error_name = "settings"
+                    self.add_error(error_name, error_detail)
+                    # Serializer might add more data to the JSON dict
+                    if settings := self.serializer.data.get("settings"):
+                        self.cleaned_data["settings"] = settings
+        else:
+            # Serializer might add more data to the JSON dict
+            if settings := self.serializer.validated_data.get("settings"):
+                self.cleaned_data["settings"] = settings
```
should maybe flip on the logic here, have valid state, first, and not valid under else

stveit commented on src/argus_htmx/destinations/forms.py:
```diff
@@ -0,0 +1,79 @@
+    settings = forms.CharField(required=True)
+
+    def __init__(self, *args, **kwargs):
+        self.request = kwargs.pop("request", None)
```
Add comment stating this is needed because the serializer needs access to the request object

stveit commented on src/argus_htmx/destinations/forms.py:
```diff
@@ -71,6 +71,47 @@ def _validate_serializer(self):
 
 
+class DestinationFormUpdate(DestinationFormCreate):
+    def __init__(self, *args, **kwargs):
+        if instance := kwargs.get("instance"):
+            settings_key = _get_settings_key_for_media(instance.media)
+            instance.settings = instance.settings.get(settings_key)
+        super().__init__(*args, **kwargs)
```
add comment explaining that this is done to convert the settings value from the whole dict to just the email/phone number. That way we dont get the entire dict in the frontend, and just the value we are interested in. This is opposite to what is done in the create form, where we translate from the value itself to a whole dict

stveit commented on src/argus_htmx/destinations/views.py:
```diff
@@ -0,0 +1,106 @@
+def _render_destinations(
+    request,
+    create_form: Optional[DestinationFormCreate] = None,
+    update_forms: list[DestinationFormUpdate] = None,
```
```suggestion
    update_forms: Sequence[DestinationFormUpdate] = None,
```

hmpf commented on src/argus_htmx/destinations/forms.py:
```diff
@@ -0,0 +1,120 @@
+        )
+
+
+def _get_settings_key_for_media(media: Media) -> str:
```
As is, assumes that there is only one setting a user can set per plugin, which currently holds but might not in the future.

hmpf commented on src/argus_htmx/destinations/forms.py:
```diff
@@ -0,0 +1,120 @@
+        # self.serializer should be initiated and validated in clean() before save() is called
+        self.serializer.save(user=self.request.user)
+
+    def _init_serializer(self):
+        serializer = RequestDestinationConfigSerializer(
+            data={
+                "media": self.cleaned_data["media"],
+                "label": self.cleaned_data.get("label", ""),
+                "settings": self.cleaned_data["settings"],
+            },
+            context={"request": self.request},
+        )
+        self.serializer = serializer
+
+    def _validate_serializer(self):
+        media = self.cleaned_data["media"]
+        settings_key = _get_settings_key_for_media(media)
+
+        # Add error messages from serializer to form
+        if not self.serializer.is_valid():
+            for error_name, error_detail in self.serializer.errors.items():
+                if error_name in ["media", "label", settings_key]:
+                    if error_name == settings_key:
+                        error_name = "settings"
+                    self.add_error(error_name, error_detail)
+                    # Serializer might add more data to the JSON dict
+                    if settings := self.serializer.data.get("settings"):
+                        self.cleaned_data["settings"] = settings
+        else:
+            # Serializer might add more data to the JSON dict
+            if settings := self.serializer.validated_data.get("settings"):
+                self.cleaned_data["settings"] = settings
+
+        if label := self.cleaned_data["label"]:
+            filter = DestinationConfig.objects.filter(label=label)
+            if self.instance:
+                filter = filter.exclude(pk=self.instance.pk)
+            if filter.exists():
+                self.add_error("label", "Name must be unique per media")
```
Hopefully all of this will go away when the plugins are changed.

stveit replied:
:crossed_fingers:

hmpf commented on src/argus_htmx/templates/htmx/destinations/destinations.html:
Should be named destination/destination_list.html

hmpf commented on src/argus_htmx/templates/htmx/destinations/edit_or_delete_form.html:
Should be named "destination/_edit_or_delete_form.html"

hmpf replied:
.. and split it into three files. On to hold em, and one per action:

- destination/_forms.html
- destination/_edit_form.html
- destination/_delete_form.html

hmpf commented on src/argus_htmx/destinations/forms.py:
Rename "argus_htmx/destination*_s_*/" to the singular: "argus_htmx/destination/"

hmpf reviewed (changes requested):
I've tested it and it works except it's a bit confusing that when I edit/delete the url changes. Maybe redirect to "/destination/" after change? @podliashanyk might have a better idea.

Almost only renames following.

hmpf commented:
Ehh, and I almost forgot: media-field of the creation-form is bugged in "argus"-theme:
![image](https://github.com/user-attachments/assets/a15b8a6b-044f-48d8-bda8-d53926b3c0f2)

stveit commented:
> I've tested it and it works except it's a bit confusing that when I edit/delete the url changes. Maybe redirect to "/destination/" after change? @podliashanyk might have a better idea.
> 
> Almost only renames following.

I imagine HTMXifying this properly should fix this. Instead of loading the entire page when updating, only the update form you used is actually updated etc

hmpf commented:
Moved to argus-server
